bite-size-notes/week-2-b-CBFs.md

```diff
@@ -0,0 +1,116 @@
+
+$$
+\text{sup}_{u\in U} [L_f h(x) + L_g h(x) u] \geq - \alpha (h(x)) \hspace{0.2cm}
+$$
+
+for all $x \in D$ and where $L_f = \nabla h(x)^T f(x)$ and $L_g = \nabla h(x)^T g(x)$.
+
+So therefore to **guarentee safety via CBFs** we consider the set consisting of all the control inputs that render the set $\mathcal C$ safe: 
+
+$$
+K_{cbf} (x) = \{u \in U : L_f h(x) + L_g h(x) u + \alpha (h(x)) \geq 0 \}.
+$$
+
+Therefore, we can clearly define the set of all control inputs at some $x \in D$ that keep the system safe. It follows that the existence of a control barrier function implies that the control system is safe. It's noted in [4] that CBFs do indeed provide the strongest possible conditions for safety since they are neccessary and sufficient given the assumptions we have on $\mathcal C$. 
+
+## Optimization-based Control with CBFs 
+Given that we have covered the theoretical definitions of control barrier functions and the fact that they provide the conditions to guarentee safety of our system, we can now explore how to incorporate CBFs into an optimization-based controller. This is called synthesizing controllers for **safety-critical control**. 
+
+Suppose we are given a nominal feedback controller $\pi(x)$ for the control affine system. This controller may not satisfy the CBF conditions for safety and thus generates control inputs $u$ that aren't in our safe set, i.e., $u \notin K$. We wish to devise a controller that modifies the nominal feedback controller in a minimally invasive fashion (as to still reach some goal state) in order to produce safe controls. Thus, we  consider a quadratic program (QP) formulation that finds a minimal perturbation on $u$:
+
+$$
+\begin{aligned}
+u(x) = \text{argmin}_{u \in U} \hspace{0.3cm} \frac{1}{2} || u - \pi(x) ||_2^2 \\
+\\
+\hspace{0.5cm} \text{s.t.} \hspace{0.3cm} L_f h(x) + L_g h(x) u \geq \alpha (h(x)) 
+\end{aligned}
+$$
+
+This gives straightforward method to devleop a controller to guarentee a system be safe. Since the objective is a min-norm and the constraint is linear in $u$, this program is a convex quadratic program that can be solved efficiently online. In practice, this QP-based controller serves as a safety filter on the nominal controller to ensure safe control inputs into the dynamics, rendering the system safe.
+
+
+## Example 
+Below is an image of a high-order CBF, but it shows the basics of having an initial trajecotry from a nominal policy approaching an obstacle and the CBF controller modifying the control inputs to steer around the obsctacle. 
+
+![cbf image](figs/cbf_pic.png)
+
+*Image from HOCBF paper: https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=9516971*
+
+Additionally here is a video example from Caltech of them implementing a CBF on a segway that shows its capabilities: https://www.youtube.com/watch?v=RYXcGTo8Chg 
+
```

Comment:
The paper Higher Order Control Barrier Function is linked directly where you left this comment... I also mention this paper in the "Current Research" section when discussing the difficulties of devising CBFs and their drawbacks. I can also mention it here? Is there something more specific you wanted me to mention about the HOCBF paper?

Reply:
Sorry for the confusion. Because I only saw a very intuitive video in your example, I figured you could add more description on how and why the example would work and relate to the paper you mentioned here. I didn't relate this paper to the paper you mentioned in the current research section... I guess what I meant was the more technical explanation of the example.

Reply:
Ah yes I see. Okay I'll relate the video (that's very easy to understand to the audience) to some further technical details explaining how it works in the video. And I'll mention the HOCBF paper. Thanks.